lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -825,12 +825,30 @@ struct DynExtractOpConversion : public OpConversionPattern<DynExtractOp> {
                                      op->getLoc());
 
       if (isa<hw::ArrayType>(resultType)) {
+        // FIXME: for packed arrays, the part that is OOB must be 0 or X and the
+        // part overlapping with the range must match the values of the range;
+        // for unpacked arrays, if part of the accessed slice is OOB the entire
+        // result is 0 or X
         rewriter.replaceOpWithNewOp<hw::ArraySliceOp>(op, resultType,
                                                       adaptor.getInput(), idx);
         return success();
       }
 
-      rewriter.replaceOpWithNewOp<hw::ArrayGetOp>(op, adaptor.getInput(), idx);
+      Value size = rewriter.create<hw::ConstantOp>(
+          op.getLoc(), adaptor.getLowBit().getType(), arrType.getNumElements());
+      Value isOOB = rewriter.create<comb::ICmpOp>(
+          op.getLoc(), ICmpPredicate::uge, adaptor.getLowBit(), size);
+      Value access =
+          rewriter.create<hw::ArrayGetOp>(op.getLoc(), adaptor.getInput(), idx);
+      int64_t bw = hw::getBitWidth(arrType.getElementType());
+      if (bw < 0)
+        return failure();
+
+      Value zeroValueRaw =
+          rewriter.create<hw::ConstantOp>(op.getLoc(), APInt(bw, 0));
+      Value zeroValue = rewriter.createOrFold<hw::BitcastOp>(
+          op.getLoc(), arrType.getElementType(), zeroValueRaw);
+      rewriter.replaceOpWithNewOp<comb::MuxOp>(op, isOOB, zeroValue, access);
```

Comment:
This looks reasonable assuming that `hw::ArrayGetOp` has undefined behaviour (or undefined result) for out-of-bounds accesses. I don't remember off the top of my head whether out-of-bounds index in `ArrayGetOp` causes a potential segfault in the lowered LLVM IR… But it might be nice if `hw::ArrayGetOp` just produce an undefined result, but don't cause arbitrary undefined behaviour such as crashing the simulator. Then we can just do exactly what you did here: read the value and then discard it with a mux if it's out of bounds.

Reply:
The documentation of the HW aggregate ops is generally a bit sparse, but I'd say a OOB `hw.array_get` produces an undefined result, but no immediate UB. The HWToLLVM lowering currently does not take this into account. I'll create an issue for that.